### PR TITLE
Chat Page 

### PR DESCRIPTION
**Pull request by knanahassouna1**
#29 #30 #28 

create chat page front end and back end 
edit auth errors
add contact us front end and back end 
edit style in home page 
add favi icon 
create heroku database

### CODE REVIEW

File: src/controllers/suggestion.js [hunk 1]
```diff
@@ -0,0 +1,24 @@
+const addSuggestion = require('./../model/queries/addSuggestion');
+const  validateEmail = (email) => {
+  const re = /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
```
Comment (MohammedYehia) [inline src/controllers/suggestion.js]: the same validation in a lot of files
in your next project use some modules like joi or express validator that would make your life easier and also my life :smile:

File: src/model/database/lastMessages.sql [hunk 2]
```diff
@@ -0,0 +1,15 @@
+BEGIN ;
```
Comment (MohammedYehia) [inline src/model/database/lastMessages.sql]: this should be in db_build.sql file

File: src/model/database/lastMessages.sql [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+CREATE TABLE archivetitles(
+  id IN NOT NULL,
+  sender varchar(30) NOT NULL,
+  receiver varchar(30) NOT NULL,
+  message varchar(300) NOT NULL
+);
+
+INSERT INTO archivetitles (id, sender, receiver, message) SELECT * FROM chats WHERE id IN (SELECT MAX(id) FROM chats GROUP BY sender, receiver) And sender IN(SELECT sender FROM chats GROUP BY sender, receiver) And receiver IN(SELECT receiver FROM chats GROUP BY sender, receiver);
+
+select * from archivetitles where sender=$1 or receiver=$1 order by id;
```
Comment (MohammedYehia) [inline src/model/database/lastMessages.sql]: what is this?

File: src/model/database/lastMessages.sql [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+DROP TABLE  IF EXISTS archivetitles CASCADE;
+
+CREATE TABLE archivetitles(
+  id IN NOT NULL,
+  sender varchar(30) NOT NULL,
+  receiver varchar(30) NOT NULL,
+  message varchar(300) NOT NULL
+);
+
+INSERT INTO archivetitles (id, sender, receiver, message) SELECT * FROM chats WHERE id IN (SELECT MAX(id) FROM chats GROUP BY sender, receiver) And sender IN(SELECT sender FROM chats GROUP BY sender, receiver) And receiver IN(SELECT receiver FROM chats GROUP BY sender, receiver);
```
Comment (MohammedYehia) [inline src/model/database/lastMessages.sql]: move this to demo.sql

File: src/model/queries/lastMessages.js [hunk 5]
```diff
@@ -0,0 +1,40 @@
+const fs = require('fs');
```
Comment (MohammedYehia) [inline src/model/queries/lastMessages.js]: why do we need this file??

File: package.json [hunk 6]
```diff
@@ -19,9 +19,10 @@
     "express-handlebars": "^3.0.0",
     "pg": "^7.4.3",
     "env2": "^2.2.2",
-    "bcryptjs":"*",
-    "jsonwebtoken":"*",
-    "cookie-parser":"*"
+    "bcryptjs": "*",
```
[the remark this answers is not in the document]
Comment (knanahassouna1) [reply package.json]: okay  i will fix it , thanks for your notes but the problem i was can't  install modules from terminal

File: public/css/admin.css [hunk 7]
```diff
@@ -1,6 +1,5 @@
-@import url('https://fonts.googleapis.com/css?family=Lato');
-@import url('https://fonts.googleapis.com/css?family=Oswald:200');
 .admin .container{
+min-height: 80vh;
```
[the remark this answers is not in the document]
Comment (knanahassouna1) [reply public/css/admin.css]: okay , hhh

File: public/js/chat.js [hunk 8]
```diff
@@ -0,0 +1,231 @@
...
+  messageDirection.appendChild(text);
+  message.appendChild(messageDirection);
+  messenger_space.appendChild(message);
+  // For Scroll Down When Add New Message
+  messenger_space.scrollTop = messenger_space.scrollHeight;
+}
+
+// Send message to sockets network
+input.addEventListener('keyup', (event) => {
+  if (event.keyCode === 13 && event.target.value.trim()) {
```
[the remark this answers is not in the document]
Comment (knanahassouna1) [reply public/js/chat.js]: okay , and thanks for article

File: public/js/chat.js [hunk 9]
```diff
@@ -0,0 +1,231 @@
...
+});
+
+// change the reciver ;
+const newMessengerConfig = (jwt, username) => {
+  messenger.setAttribute('username', username);
+  messenger.setAttribute('token', jwt);
+  const messenger_space = messenger.getElementsByClassName('messenger_space');
+  const usernameSpan = messenger.getElementsByTagName('span')[0];
+  usernameSpan.textContent = username;
+  messenger_space[0].innerHTML = '';
```
[the remark this answers is not in the document]
Comment (knanahassouna1) [reply public/js/chat.js]: okay

File: src/model/database/lastMessages.sql (hunk 3 above)
Comment (knanahassouna1) [reply src/model/database/lastMessages.sql]: that's query to get last message for all users send message to the current user.

File: src/model/queries/lastMessages.js (hunk 5 above)
Comment (knanahassouna1) [reply src/model/queries/lastMessages.js]: i was try something and forget remove these modules

File: src/model/database/lastMessages.sql (hunk 4 above)
Comment (knanahassouna1) [reply src/model/database/lastMessages.sql]: is not a demo is a part of the query

**Pull request merged**